statsmodels/base/distributed_estimation.py

```diff
@@ -0,0 +1,661 @@
+
+    Returns
+    -------
+    A tuple of parameters for regularized fit
+        An array-like object of the fitted parameters, params
+        An array-like object for the gradient
+        A list of array like objects for nodewise_row
+        A list of array like objects for nodewise_weight
+    """
+
+    score_kwds = {} if score_kwds is None else score_kwds
+    hess_kwds = {} if hess_kwds is None else hess_kwds
+
+    if fit_kwds is None:
+        raise ValueError("_est_regularized_debiased currently " +
+                         "requires that fit_kwds not be None.")
+    else:
+        alpha = fit_kwds["alpha"]
+
+    if "L1_wt" in fit_kwds:
+        L1_wt = fit_kwds["L1_wt"]
+    else:
+        L1_wt = 1
+
+    nobs, p = mod.exog.shape
+    p_part = int(np.ceil((1. * p) / partitions))
+
+    params = mod.fit_regularized(**fit_kwds).params
+    grad = _calc_grad(mod, params, alpha, L1_wt, score_kwds) / nobs
+
+    wexog = _calc_wdesign_mat(mod, params, hess_kwds)
+
+    nodewise_row_l = []
+    nodewise_weight_l = []
+    for idx in range(mnum * p_part, min((mnum + 1) * p_part, p)):
+
+        nodewise_row = _calc_nodewise_row(wexog, idx, alpha)
+        nodewise_row_l.append(nodewise_row)
+
+        nodewise_weight = _calc_nodewise_weight(wexog, nodewise_row, idx,
+                                                alpha)
+        nodewise_weight_l.append(nodewise_weight)
+
+    return params, grad, nodewise_row_l, nodewise_weight_l
+
+
+def _join_debiased(results_l, partitions, threshold=0):
+    """joins the results from each run of _est_regularized_debiased
+    and returns the debiased estimate of the coefficients
+
+    Parameters
+    ----------
+    results_l : list
+        A list of tuples each one containing the params, grad,
+        nodewise_row and nodewise_weight values for each partition.
+    partitions : scalar
+        The number of partitions that the data will be split into.
+    threshold : scalar
+        The threshold at which the coefficients will be cut.
+    """
+
+    p = len(results_l[0][0])
+
+    params_mn = np.zeros(p)
+    grad_mn = np.zeros(p)
+
+    nodewise_row_l = []
+    nodewise_weight_l = []
+
+    for r in results_l:
+
+        params_mn += r[0]
+        grad_mn += r[1]
+
+        nodewise_row_l.extend(r[2])
+        nodewise_weight_l.extend(r[3])
+
+    params_mn /= partitions
+    grad_mn *= -1. / partitions
+
+    approx_inv_cov = _calc_approx_inv_cov(nodewise_row_l, nodewise_weight_l, p)
+
+    debiased_params = params_mn + approx_inv_cov.dot(grad_mn)
+
+    debiased_params[np.abs(debiased_params) < threshold] = 0
+
+    return debiased_params
+
+
+class DistributedModel():
+    __doc__ = """
+    Distributed model class
+
+    Parameters
+    ----------
+    data_generator : generator
+        A generator that produces a sequence of tuples where the first
+        element in the tuple corresponds to an endog array and the
+        element corresponds to an exog array.
+    partitions : scalar
```

Comment:
Each machine needs to know how many partitions there are so that we can split up the variables for the node-wise regression.  So not in general, but for the debiased regularized case we need something like this.  This might be one argument for allowing for a more general class of generators, because we could split the variables up in the generator instead of handling them internally.